### DISCUSSION TITLE
App isn't responding

### DISCUSSION THREAD

**Discussion by Jay-57blocks** [dotnet/maui#25523] [Q&A] [answered]
### Description

Hi，

My project occasionally encounters unresponsive system popups, and I was wondering if maui has any tools to help troubleshoot this issue? For example, tools to analyze CPU usage, memory usage, etc.

![Image](https://github.com/user-attachments/assets/c9b7ff16-4706-415a-8eee-a9bc051eef8d)
![Image](https://github.com/user-attachments/assets/f9c73591-3008-4bed-89d9-817e22b2bb98)

### Steps to Reproduce

This is an occasional problem with no reproduction steps

### Version with bug

8.0.61 SR6.1

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Affected platforms

Android

### Affected platform versions

安卓14

### Did you find any workaround?

I would like to know how to locate and analyze this issue, sorry I am not familiar with maui.

I hope to get your help.

**Comment by bhavanesh2001** [comment]
@Jay-57blocks  You use NavigationPage for navigation  in app?

**Reply by Jay-57blocks** [reply, marked as answer]
Yes, I use Prism.Navigation for navigation.

```
<PackageReference Include="Prism.DryIoc.Maui" Version="9.0.218-pre" />
<PackageReference Include="Microsoft.Maui.Controls" Version="8.0.40" />
<PackageReference Include="Microsoft.Maui.Controls.Compatibility" Version="8.0.40" />
```

**Reply by Jay-57blocks** [reply]
Hi bhavanesh2001，
I would like to know how to troubleshoot this issue, or if the need to upgrade maui to the latest version would reduce the probability of ANR popup boxes?

**Comment by bhavanesh2001** [comment]
@Jay-57blocks  I faced similar issue a while back.... Eventually I found this [plugin](https://github.com/AdamEssenmacher/MemoryToolkit.Maui). Read the Readme of the plugin you may get an understanding of what's happening.

The simple reason is  "**Memory Leaks**" . You can use Android Studio profiler to track you app memory allocation while debugging. In my case the app's memory kept on increasing(while using app) to point where system throws that ANR popup boxes.

Note: The plugin has it's own drawbacks. Refer this [thread](https://github.com/AdamEssenmacher/MemoryToolkit.Maui/issues/23) by me.

**Reply by Jay-57blocks** [reply]
Hi bhavanesh2001，

Thank you very much for your help, this problem has been bothering me for a long time


I checked the CPU and memory usage using "Android Studio profiler" and it looks like the RAM usage is a little bit high, my android phone has only 3GB of RAM, may I ask what percentage of RAM reaches the phone's RAM in general that triggers ANR in your case?
For example, 50% of 3GB. When the App memory footprint is greater than 1.5GB, it may cause the system to garbage collect or degrade performance, thus increasing the risk of ANR.

<img width="1892" alt="image" src="https://github.com/user-attachments/assets/0dcf3095-2626-472e-8309-105a444168ec">


Also, I remember when I was experiencing ANR, the console logs alerted me that there were too many tasks on the CPU, and that my case might not be caused by memory alone
![image](https://github.com/user-attachments/assets/d133ec63-47df-4496-afe7-d745bc4bb7ab)

**Reply by bhavanesh2001** [reply]
@Jay-57blocks  I guess this differs from device to device. In my case, for a device with 6GB RAM when my app usage reaches around 650-700Mb the ANR popups start showing.

**Comment by Jay-57blocks** [comment]
Hi，

The latest news, I finally found the reason, is because my colleague's part
of the code caused a thread deadlock. Finally found the reason.


Best regards/Jay Ren


Bhavanesh N ***@***.***>于2024年11月4日 周一17:19写道：
